File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/py3/test_all.py
# meta_puzzles by Sebastien Rubens
#
# Please go to https://github.com/seb-pg/meta_puzzles/README.md
# for more information
#
# To the extent possible under law, the person who associated CC0 with
# meta_puzzles has waived all copyright and related or neighboring rights
# to meta_puzzles.
#
# You should have received a copy of the CC0 legalcode along with this
# work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

def get_local_modules():
    import os
    import re
    import importlib.util
    dirname = os.path.dirname(__file__)
    w = os.walk(dirname)
    files = next(w)[2]
    pattern = re.compile("^l\d_.*.py$")
    #files = sorted(os.sep.join([dirname, f]) for f in files if pattern.search(f) is not None)
    files = sorted(f[:-3] for f in files if pattern.search(f) is not None)
    tests_fns = []
    for f in files:
        m = importlib.import_module(f)
        if m:
            fn = getattr(m, 'tests', None)
            if fn:
                tests_fns.append((f, fn))
    return tests_fns


def format_args(args):
    return


if __name__ == "__main__":
    tests_fns = get_local_modules()
    nb_errors = 0
    for tests_name, tests_fn in tests_fns:
        ret = tests_fn()
        test_fn, extend_fn, all_tests = ret[:3]
        cmp = lambda res, exp: res == exp
        if len(ret) > 3:
            cmp = ret[3]
        for sub_name, tests in all_tests:
            print()
            print(tests_name, sub_name)
            if not tests:
                print("  no tests")
                continue
            for nb, (args, expected) in enumerate(tests, 1):
                ex_args = extend_fn(*args)
                res = test_fn(*ex_args)
                if cmp(res, expected):
                    print("  %s, test #%s: res=%s CORRECT" % (sub_name, nb, res))
                else:
                    print("  %s, test #%s: res=%s expected=%s ERROR <-----" % (sub_name, nb, res, expected))
                    nb_errors += 1
    print("\n%s errors found" % nb_errors)


# End

File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/cs/TestAll.cs
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

using System;
using System.Collections.Generic;

namespace test_all
{

    class Res<T>
    {
        public T res;
    }

#nullable enable annotations

    class TestAll
    {

        public static int run_all_tests<Args, Ret>(String name, List<Args> args_list, Func<Args, Ret> fn, double precision = 0.0)
            where Args : Res<Ret>
            where Ret : IEquatable<Ret>
        {
            Console.WriteLine("\n{0}", name);

            int nb_errors = 0;
            int nb = 1;
            foreach (Args args in args_list)
            {
                var res = fn(args);
                bool is_same = res.Equals(args.res);
                if (typeof(Ret) == typeof(double))
                {
                    var x = (double)Convert.ChangeType(res, typeof(double));
                    var y = (double)Convert.ChangeType(args.res, typeof(double));
                    is_same = Math.Abs(x - y) <= precision;
                }

                if (is_same)
                    Console.WriteLine("  test #{0}: res={1} CORRECT", nb, res);
                else
                {
                    Console.WriteLine("  test #{0}: res={1} expected={2} ERROR <-----", nb, res, args.res);
                    nb_errors += 1;
                }
                ++nb;
            }

            return nb_errors;
        }

    }

}
